### ISSUE TITLE
optimizer: allow self-join without hash exchange

### ISSUE THREAD

**Issue by BugenZhao** [risingwavelabs/risingwave#6233]
The distribution of `StreamTableScan` is `UpstreamHashShard`, where we're unknown of the distribution since different tables may use different consistent hash vnode mapping.

With #6098, we now have the `table_id` in the `UpstreamHashShard`. If we're doing a self-join, we can avoid the hash exchange before them and make the output `UpstreamHashShard` as well. cc @chenzl25 

https://github.com/risingwavelabs/risingwave/blob/a9d668ccb0a42379c4c672f614b2f2691e91cf5d/src/frontend/src/optimizer/plan_node/logical_join.rs#L897-L918

```
dev=> explain create materialized view mv as select t1.v from t t1, t t2 where t1.v = t2.v;
                                                              QUERY PLAN                                                               
---------------------------------------------------------------------------------------------------------------------------------------
 StreamMaterialize { columns: [v, t._row_id(hidden), t._row_id#1(hidden), t.v(hidden)], pk_columns: [t._row_id, t._row_id#1, v, t.v] }
 └─StreamHashJoin { type: Inner, predicate: t.v = t.v }
   ├─StreamExchange { dist: HashShard(t.v) }
   | └─StreamTableScan { table: t, columns: [v, _row_id] }
   └─StreamExchange { dist: HashShard(t.v) }
     └─StreamTableScan { table: t, columns: [v, _row_id] }
(6 rows)
```

Same for batch join.

**Comment by chenzl25** [comment]
Yes. This optimization will cause a fragment contains more than one chain operator and we need to take care.

**Comment by BugenZhao** [comment]
> a fragment contains more than one chain operator and we need to take care

Yes. I believe there will some problems with the current MV on MV creation reporting since we're using the actor id. 😢

**Comment by BugenZhao** [comment]
Resolved by #9320
